Adding some tests to System.Linq.

rafidka opened the pull request:
Currently, only limited number of tests to get feedback before proceeding with adding more tests.

#1143

dnfclas commented:
Hi __@rafidka__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

ellismg commented:
The File System Watcher tests all seemed to pass, but xunit returned an error:

http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_debug_prtest/430/console

```
d:\j\workspace\dotnet_corefx_windows_debug_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00018\lib\tests.targets(67,5): warning MSB3073: The command "corerun.exe xunit.console.netcore.exe System.IO.FileSystem.Watcher.Tests.dll  -xml .\testResults.xml -notrait category=failing " exited with code -1073741819. [d:\j\workspace\dotnet_corefx_windows_debug_prtest\src\System.IO.FileSystem.Watcher\tests\System.IO.FileSystem.Watcher.Tests.csproj]
```

Only happened on one leg.

ellismg commented:
@dotnet-bot test this please

rafidka commented:
@ellismg from the log, it doesn't look like this is related to my change, or am I missing something?

  Discovering: System.Linq.Tests
  Discovered:  System.Linq.Tests
  Starting:    System.Linq.Tests
  Finished:    System.Linq.Tests
  
  === TEST EXECUTION SUMMARY ===
     System.Linq.Tests  Total: 13, Errors: 0, Failed: 0, Skipped: 0, Time: 0.197s
  XmlDiff -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\XmlDiff\XmlDiff.dll
  System.ComponentModel.Primitives.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.ComponentModel.Primitives.Tests\System.ComponentModel.Primitives.Tests.dll
  System.ComponentModel.EventBasedAsync.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.ComponentModel.EventBasedAsync.Tests\System.ComponentModel.EventBasedAsync.Tests.dll
  System.Console.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Console.Tests\System.Console.Tests.dll
  System.Reflection.Metadata.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Reflection.Metadata.Tests\System.Reflection.Metadata.Tests.dll
  System.Collections.Concurrent.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Collections.Concurrent.Tests\System.Collections.Concurrent.Tests.dll
  System.IO.FileSystem.Primitives.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.IO.FileSystem.Primitives.Tests\System.IO.FileSystem.Primitives.Tests.dll
  System.Collections.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Collections.Tests\System.Collections.Tests.dll
  System.Xml.XPath.XmlDocument.Tests -> d:\j\workspace\dotnet_corefx_windows_debug_prtest\bin\Windows_NT.AnyCPU.Debug\System.Xml.XPath.XmlDocument.Tests\System.Xml.XPath.XmlDocument.Tests.dll
  xUnit.net console test runner (32-bit .NET Core)
  xUnit.net console test runner (32-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
  
  Copyright (C) 2014 Outercurve Foundation.

rafidka commented:
@Maxwe11, yeah, true, I should have used loops (which I used in other places to be honest), but I just wanted to push things quickly to get some feedback. I will change that. I will also change the assertion statements as required to make sure the expected value is passed correctly.

stephentoub commented:
@dotnet-bot test this please

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,131 @@
+    public partial class EnumerableTests
+    {
+        [Fact]
+        public void Average()
```
Min?

rafidka replied:
It should be average actually, but I didn't update the code. I will fix that.

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,131 @@
+            var array = Enumerable.Range(1, 10).ToArray();
+            var min = array.Min();
+            Assert.Equal(min, 1);
+        }
```
Presumably there will be other tests for Min, e.g. with an empty sequence, with a sequence with only one value, with negative values, with an enumerator that throws an exception, etc.  Do you plan to do all of those in a single Min() test, or do you plan to split them across multiple Min_SomeSpecificTest() tests?  I don't have a strong preference either way, but if your plan is the latter, you probably want to start with that naming scheme now.

rafidka replied:
Yes, indeed, just wanted to get an initial feedback from you guys.
As for the plan, I am planning to write a generic code for testing that takes 1) an input 2) a lambda expression of how to verify the correctness of the code, and then I will pass multiple input types. This way I will avoid to write the code multiple times, yet I can ensure I cover different inputs with different types. Does that make sense?

stephentoub replied:
In some cases that may be ok.  In other cases it might be overly complicated (e.g. the above example can be simply done in a single statement as ```Assert.Equals(1, Enumerable.Range(1, 10).Min());```... that's easy to read, and there's no need for the complication of lambdas here).  And for some cases it'll be insufficient (e.g. negative tests cases where you want to verify that an exception occurs).

Clockwork-Muse replied:
Note also that lambdas don't print out nicely, so you may have a hard time telling which one failed...

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,131 @@
+        [Fact]
+        public void Average()
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
```
Why ".ToArray()"?

rafidka replied:
Well, there is no reason to go do ToArray() apart from avoiding to enumerate the enumerable multiple times. OK, I will modify the tests such that they test different input types.

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,131 @@
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
+            Assert.True(array.All(i => i > 0));
+            Assert.False(array.All(i => i > 1));
```
I don't think you actually need to do this 10 times; it should be sufficient to test a true and a false case.  If you really wanted to do it for all of the values, though, it'd be cleaner to have a for loop around a single Assert.False rather than duplicating it for each value.

rafidka replied:
Yeah, sure, I will use a loop.

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,131 @@
+        {
+            var array = Enumerable.Range(1, 10).ToArray();
+            var min = array.Min();
+            Assert.Equal(min, 1);
```
Please make sure that the values are passed to the right arguments to Assert's methods, e.g. the first parameter to Equal is "expected" and the second parameter is "actual", but here you've reversed them.  While this will still trigger an assertion on failure, the output message will be confusing because the values will be reversed.

dnfclas commented:
__@rafidka__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

Maxwe11 commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+
+            // Ensure that every element in 'array' exists in 'distinctList'.
+            foreach (var i in originalList)
```
Consider
`Asser.DoesNotContain(originalList, x => !distinctList.Contains(x));`

rafidka replied:
Thanks @Maxwe11, yeah, this is better.

Maxwe11 commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+
+            // Ensure that every element in 'distinctList' exists in 'originalList'.
+            foreach (var i in distinctList)
```
Consider
`Asser.DoesNotContain(distinctList, x => !originalList.Contains(x));`

rafidka replied:
Thanks.

Clockwork-Muse replied:
149 +            // Ensure that every element in 'array' exists in 'distinctList'. 
    150 +            foreach (var i in originalList) 
    151 +            { 
    152 +                var correct = false; 
    153 +                foreach (var j in distinctList) 
    154 +                { 
    155 +                    if (i.Equals(j)) 
    156 +                        correct = true; 
    157 +                } 
    158 +                Assert.True(correct, "Some elements in the original enumerable are missing in the result of Distinct()."); 
    159 +            } 

Yeah, no, lets not do tests this way please.  I've been cleaning up this sort of thing from the PLINQ tests.

You should be using:

    foreach(var i in originalList)
    {
        Assert.Contains(i, distinctList);
    }

Or even:

    Assert.All(originalList, i => Assert.Contains(i, distinctList));

...which will even print out the index of the offending element, although it means the output wouldn't be pipelined (because internally it calls `ToArray()`, although you shouldn't rely on that particular behavior)...
You are planning on pipelined vs non-pipelined output tests, aren't you?

Side note - despite the comments in that test, you _aren't_ sorting it first (it assumes sorted input).  It would be better to use some sort of set for this verification (either grab `HashSet`, or you can make a trivial helper set class in about 10 lines).

rafidka replied:
Sure, I will do that.

rafidka commented:
@stephentoub, I see that this PR is marked as All is well, but it hasn't been merged yet. I am still new to contributing here, so I don't know what I should do now. Should I wait until this PR is merged, so that I start a new one with new changes? Or should I keep adding to this one?

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+            Assert.False(array.Any(i => i > 10));
+        }
+
+        private void FindCountAndValidate<T>(IEnumerable<T> enumerable, int expectedCount)
```
I don't understand the name of this method... what's the "Find" part meant to convey?

rafidka replied:
So to avoid repeating the logic multiple times, I am defining a method that finds the count of an IEnumerable of any type and then validate that the count is correct. Then, as you can see below, I am calling this multiple times with enumerables of different data types.

As for the Find part, it is just a way of saying this method is going to Find the count.

Does that make sense?

stephentoub replied:
I see.  Since "Count" is itself a verb, I think "CountAndValidate" is fine.

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+            Assert.Equal(expectedCount, enumerable.ToList().Count());
+            Assert.Equal(expectedCount, enumerable.ToArray().Count());
+            Assert.Equal(expectedCount, new Stack<T>(enumerable).Count());
+            Assert.Equal(expectedCount, new Queue<T>(enumerable).Count());
```
You probably don't need both Stack and Queue; one of them would be sufficient to cover the case of ICollection.

rafidka replied:
Would you like me to remove one of them?

stephentoub replied:
Yes, please.  We want to be comprehensive, but we should also be thoughtful about the value of each test.  It's great that you're covering enumerables, lists (```ICollection<T>```), arrays, and ```ICollection```s, but we probably don't need to double up on any of them, unless there's some specific difference between ```Stack<T>``` and ```Queue<T>``` you were trying to test here?

rafidka replied:
No, there wasn't any reason, more than I just wanted to test all different collection types without making any assumption about how they are implemented. OK, I will remove one of them.

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+            FindCountAndValidate(range.Select<int, decimal>(i => i), count);
+        }
+
+        private void FindDistinctAndValidate<T>(IEnumerable<T> original)
```
Ditto on "Find"

rafidka replied:
So should I just rename it to CountAndValidate()?

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+            // Ensure the result doesn't contain duplicates; sort it first to make this easier.
+            // Notice that we assume the correctness of OrderBy() here, which is tested
+            // in another unit.
+            for (var i = 0; i < distinctList.Count - 1; i++)
```
Some of this method's implementation could be simplified with HashSet.

stephentoub commented on src/System.Linq/tests/EnumerableTests.cs:
```diff
@@ -0,0 +1,237 @@
+                    if (i.Equals(j))
+                        correct = true;
+                }
+                Assert.True(correct, "Some elements in the result of Distinct() don't xist in the given enumerable.");
```
Nit: typoe: 'xist' -> 'exist'

rafidka replied:
Thanks, will fix that.

stephentoub commented:
> I see that this PR is marked as All is well, but it hasn't been merged yet.

That just means that the continuous integration build system verified that everything compiled and all tests passed.  Any time you submit a PR or update a PR, that'll run.

> Should I wait until this PR is merged, so that I start a new one with new changes? Or should I keep adding to this one?

Are you done with this?  I see comments like "Sure, I will address that" that haven't been addressed, e.g. @Clockwork-Muse's comments about how to write asserts, so I wasn't sure what was the status.

> Do you want me to make the tweak before the PR can be merged? I prefer to do it later, as I really want to avoid having a large code which is not in sync with the main repository.

Ideally we wouldn't be checking in things that we know need to be fixed.  If you really want to hold off, ok, but I didn't understand the justification of this not being in sync with the main repo... for one, no one else is currently modifying these tests, so there shouldn't be any conflicts, but if you were really concerned, you can easily keep your branch in sync, e.g. by rebasing on the current master.

> Or should I keep adding to this one?

Once any issues in the PR are addressed, we can go ahead and merge it.  I realize here you were just looking to get feedback on the approach, and that's great.  In the future, it'd be very good to be more methodical about what's being tested.  For example, I didn't know whether to comment on things that were missing because I wasn't sure what the intent was, e.g. negative test cases, empty enumerables, overflow tests, etc.

Thanks!

rafidka commented:
> Are you done with this? I see comments like "Sure, I will address that" that haven't been addressed, e.g. @Clockwork-Muse's comments about how to write asserts, so I wasn't sure what was the status.

Well, I thought I would wait until it is merged so that I can get it to merged before my branch gets really outdated. But following your explanation above, I don't see any harm in fixing the addressed issues and then making another submission.

Thanks for the comment, I will address the issues and commit the changes.

rafidka commented:
@stephentoub, I believe I made all the changes the community asked to do. Please let me know if there are still more changes you want me to make.

rafidka commented:
I just noticed the build has failed, but I don't know to be honest. I tested the changes locally and they do work, and the details page doesn't show up anything:

> No problems were identified. If you know why this problem occurred, please add a suitable Cause for it.

Any idea?

akoeplinger commented:
@rafidka There's an error in the [console log](http://dotnet-ci.cloudapp.net/job/dotnet_corefx_windows_release_prtest/514/console) for one of the jobs (you can find it by going to the [log](http://dotnet-ci.cloudapp.net/job/dotnet_corefx_prtest/525/console) for the PR job):

```
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00018\lib\tests.targets(67,5): warning MSB3073: The command "corerun.exe xunit.console.netcore.exe System.IO.FileSystem.Watcher.Tests.dll  -xml .\testResults.xml -notrait category=failing " exited with code -1073741819. [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem.Watcher\tests\System.IO.FileSystem.Watcher.Tests.csproj]
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00018\lib\tests.targets(73,5): error : One or more tests failed while running tests from 'System.IO.FileSystem.Watcher.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.IO.FileSystem.Watcher\tests\System.IO.FileSystem.Watcher.Tests.csproj]
```

... but I don't think that's related to your PR.

stephentoub commented:
Yeah, it's not related to this PR; #1157

stephentoub commented:
@dotnet-bot test this please

stephentoub commented:
Thanks for cleaning this up, @rafidka.  I left a few more minor comments, but they can be addressed separately when you add more tests.  I'll go ahead and merge.  Thanks!

rafidka commented:
Thanks @stephentoub. I will address your comments. Happy that my changes finally made their way to corefx repository :-)
